Changes to support running slackroll under either Python 2 or Python 3.

spillner opened the pull request:
I took a crack at initial Python 3 support. This version of the script can be run under either Python 2 or Python 3, with compatible on-disk DB formats so you can switch back and forth at will. Note that you will need to install either the berkeleydb or bsddb3 module within your Python 3 environment (neither are part of the default Python 3 installation on slackware-current), as these are not a built-in part of Python 3 dbm. Python 3's urllib complains about the deprecated use of urlopener, but it works just fine.

I've been dogfooding this for a bit but it has not yet had comprehensive test coverage.

spillner committed:
Changes to support running slackroll under either Python 2 or Python 3.

aclemons commented:
Thanks for this. I will have a thorough look over it this weekend, but the diff is already smaller than I expected, which is great.

spillner committed:
Improvements to ensure on-disk database (especially persistent.db) cross-compatibility between Python 2 and 3, and different Berkeley DB implementations for Python 3

aclemons commented:
Ok, this has grown larger since I first looked it. I think it would be easier if we could have commit for each type of compatibility change, so it is clear what is changing and why.

aclemons commented on slackroll:
```diff
@@ -14,10 +14,47 @@
+import sys
+
+# BCS: Python 3 'dbm' == Python 2 'anydbm', both differ from Python 2's 'dbm'
+if sys.version_info[0] <= 2:
```
Can we create a constant for this (`sys.version_info[0] <= 2` ) so we only have to define it once?

spillner replied:
Addressed in b31999b.

Regarding your previous comment, I actually hadn't ever intended to push a public version with only a703d8b (minimum changes to run under either Python 2 or 3) but not e7fdb31 (changes to ensure writing to the shelve/pickle databases under Python 3 doesn't render those files later unopenable under Python 2)--- I accidentally created the PR from a branch that had only the first landed and didn't realize my error until you pointed out how short the diff was.  If you apply only a703d8b you should be perfectly happy under Python 3, but you risk a bunch of "expected value type string, got bytearray" exceptions inside pickle if you ever go back to Python 2 (without first reverting to older database files).  So while technically yes, you could separate the two, I think it's much user-friendlier to treat the two commits as a pair.  The total diff length roughly doubles, but it's still not too much to review and I think it's worth it for the convenience of being able to switch back and forth between Python 2 and 3 when testing and looking for regressions.

spillner committed:
Defined global using_python2 variable for better code clarity.

aclemons committed:
Wrap the "main" method as on master to allow merging.

This mirrors 42c9909.

aclemons committed:
Merge remote-tracking branch 'origin/master' into python3-support

spillner committed:
Prefer raw_input() over input() when on Python 2, to improve isolation against unexpected user input. Protection is automatic in Python 3.

spillner committed:
Merge branch 'python3-support' of https://github.com/spillner/slackroll into python3-support
Pulled in aclemons changes of 20-21 Jul 2022.

aclemons committed:
Merge remote-tracking branch 'origin/master' into python3-support

spillner committed:
Merge branch 'fixed-io' to resolve inconsistencies between Python 2.x and 3.x:
  - sort() now uses 3.x-style key() functions instead of 2.x-style cmp()
  - Converted file open() calls to use io.open() and specify file encoding
  - Added fallback encodings in write_to_pipe() (so non-ASCII characters
    can be sent to the pager correctly under Python 2.x)
[This is a cherry-pick of original 20d597, rebased onto v55 master.]
This represents WIP; further I/O fixes follow in original commits 778cc59 and 0deb62c

spillner committed:
Further I/O cleanup: renamed write_to_pipe() as write_to_stream(), abstracted away read_from_file()/read_lines_from_file()
[This is a cherry-pick of the original 778cc59, rebased on upstream v55 master.]
Still WIP; fully version-independent I/O also requires the original 0deb62c

spillner committed:
Cleaned up and standardized Unicode/bytestring handling between Python 2 and Python 3
[Cherry-pick of original 0deb62c, rebased onto upstream v55 master.]
This version should be fully usable in either Python 2 or Python 3.

spillner committed:
Adjusted prioritised-pkgs tests to use transient_key() instead of transient_cmp()
[Cherry-pick from original 65660b1, rebased onto upstream v55 master.]

spillner committed:
Updated test harness to use symbolic names for package states.
[Cherry-picked original 089241a, rebased on upstream v55 master.]

spillner committed:
Merge remote-tracking branch 'upstream/master' into python3-support
Reconciling this feature branch (and my fork) against v55 of upstream.
This is a WIP commit; still need to cherry-pick some of my later I/O
changes to get full functionality (e.g. output pager) under Python 3.

spillner committed:
Identifier typo corrections from old 8dd401aa.  Not strictly part of Python3 compatibility but helps resolve downstream merges.

spillner committed:
Minor Python 2/3 improvements from original commits 268a547, ce4d7ab, and 2cb019d.
- Removed the obsolete URLOpener class; it's not needed under either
  Python 2 or 3 and generates spurious warning messages under the latter.
- Guard against invoking isinstance(__, bytes) under Python 2.
- Adjusted ChangeLog parsing to ensure correct recognition of first
  duplicate entry when switching back and forth between Python 2 and 3.

spillner committed:
Merge branch 'python3-support' of github.com:spillner/slackroll into python3-support

spillner committed:
Updated GitHub workflow to use actions/cache 4.3.0, as all pre-4.2
versions have been sunset and are not longer usable in the CI pipeline.

spillner commented:
I'm not sure how many other people are still using this tool, but I figured it's silly to maintain two separate forks, and we know Python2 won't be around forever.  I rebased this on your v55, allowing slackroll to run under either python2 or python3.  If/when this is accepted, I'll open a series of separate PRs for my other feature additions (smarter version comparison, support for ChangeLogs in multiple repos, separate handling for kernel packages, and additional commands).

Looks like the CI pipeline is failing (in ruff) due to lack of type annotations for the new functions... I'll go after that tomorrow.

spillner committed:
Added type annotations and minor corrections to satisfy ruff.

spillner committed:
Final changes to pass CI pipeline lint.

spillner committed:
MINOR: Remembered why it's necessary to break this into two lines, to take advantage of the more flexible print() definition

spillner committed:
Updated new type signatures to proper union types

spillner committed:
Updated type signatures to use Either[]; | syntax not support by CI env's Python 3.9

spillner committed:
Discovered CI environment also lacks typing.Either; reverting type signature improvements until we move to Python 3.10 or later

spillner committed:
Ensure update-manifest works under Python 3 as well as 2.  Missed one read_stream() conversion in the previous rebase.
